Locali(s)(z)ed keyboard support for 3rd level characters

exsilium opened the pull request:
This patch includes support-beginnings for locali(s)(z)ation needed by countries where keyboard includes third-level characters.
Characters like @|{[]}\ are currently often not usable because they are mapped as 3rd level characters on different international keyboard layouts.
- kbLayout option introduced, which can be given during Terminal.call and changed even after, defaults to 'us';
- Placeholders for the following layouts:
  -> French (fr);
  -> German (de);
  -> UK (uk);
  -> South Slavic Latin countries - Croatia (hr),Slovenia (si), Bosnia and Herzegovina (ba), Serbia (rs);
  -> Belgium (be);
  -> Polish (pl);
  -> Turkish (tr);
- Placeholder keyboard layouts currently output "Not Implemented" to browser console;
- Estonian (ee) keyboard layout implementation for Windows (AltGr) and Mac (Alt);
- Should not break anything that was working before...

This patch will hopefully pave way for others to include their specific country layouts and solve issues like reported here: https://github.com/chjj/tty.js/issues/64

exsilium committed:
Locali(s)(z)ed keyboard support for 3rd level characters

This patch includes support-beginnings for locali(s)(z)ation needed by countries where keyboard includes third-level characters.
Characters like @|{[]}\ are currently often not usable because they are mapped as 3rd level characters on different international keyboard layouts.

- kbLayout option introduced, which can be given during Terminal.call and changed even after, defaults to 'us';
- Placeholders for the following layouts:
    -> French (fr);
    -> German (de);
    -> UK (uk);
    -> South Slavic Latin countries - Croatia (hr),Slovenia (si), Bosnia and Herzegovina (ba), Serbia (rs);
    -> Belgium (be);
    -> Polish (pl);
    -> Turkish (tr);
- Placeholder keyboard layouts currently output "Not Implemented" to browser console;
- Estonian keyboard layout implementation for Windows (AltGr) and Mac (Alt);
- Should not break anything that was working before...

rmtew commented on src/term.js:
```diff
@@ -2603,7 +2604,84 @@ Terminal.prototype.keyDown = function(ev) {
           key = '\x1b' + (ev.keyCode - 48);
         }
-      }
+      } if (this['kbLayout'] != 'us' && ((this.isMac && ev.altKey) || (!this.isMac && ev.altKey && ev.ctrlKey))) {
```
This looks like a typo, and that it should be "else if".

tonila replied:
Tested with "else if" and it did not work.

exsilium replied:
It will not work with "else if" - with the current implementation it needs to be always executed if the condition is true. Otherwise the execution stops in one of the previous else ifs. For example @ line 2599:

```
} else if ((!this.isMac && ev.altKey) || (this.isMac && ev.metaKey)) {
```

What platform, browser and which kbLayout are you testing out with, other than ee you will only receive 'Not implemented' in your browser log.

exsilium commented:
@chjj any comments on this one? It shouldn't break anything for people using default us keyboard and merging this would enable me to consume upstream :eyes: 

Other more exotic kbLayouts can be added as needed...

Serveurperso commented:
Please add Alt.Gr chars for french keyboard:(

yoshiokatsuneo commented:
I tried to handle the issue by hidden textarea (Especially Japanese for me)
https://github.com/chjj/term.js/pull/97

rgbkrk commented:
A worthy fork that seems to be moving on and welcoming your patches exists in https://github.com/sourcelair/xterm.js, maintained by @parisk.

exsilium commented:
@rgbkrk Thanks, I'll check it out!

parisk commented:
@exsilium we would be happy to work with you on this at [xterm.js](https://github.com/sourcelair/xterm.js), where we have an open issue that seems to match this PR (sourcelair/xterm.js#33).

Thanks for the mention @rgbkrk.

exsilium committed:
Revert "Locali(s)(z)ed keyboard support for 3rd level characters"

This reverts commit 6b905127980f904efeddd70b43539f32240b4a63.

exsilium committed:
Compatibility and important fixes

- 3rd level characters now work;
- Special character input via composition events work;
- Different set of fixes and improvements from xterm.js

exsilium committed:
Removed the different various modes and dead code/variables

exsilium committed:
Updated Readme and added ToDo section

exsilium committed:
Textarea id; PuTTY-style copy-paste ToDo removed

- Textarea id introduced so that the element can be addressed from the application level.
- PuTTY-style copy-paste implemented in cloud9 terminal side. This is an application level feature which is unsuitable to be done directly in term.js.

exsilium committed:
Make OS level paste work via keyboard shortcut

- Cmd-V on mac
- Ctrl-V elsewhere
- Aims to fix one of the issues reported in exsilium/cloud9#21

exsilium commented:
We've all moved over to https://github.com/xtermjs/xterm.js and this project is no longer maintained. Closing this PR and cleaning up. 📯 🚀
